core/drivers/crypto/caam/utils/utils_dmaobj.c

```diff
@@ -0,0 +1,1395 @@
+
+	if (!obj || !obj->priv)
+		return;
+
+	priv = obj->priv;
+	if (priv->type & DMAOBJ_OUTPUT)
+		op = TEE_CACHEFLUSH;
+
+	dmaobj_cache_operation(op, obj);
+}
+
+size_t caam_dmaobj_copy_to_orig(struct caamdmaobj *obj)
+{
+	struct priv_dmaobj *priv = NULL;
+	struct sgtdata *sgtdata = NULL;
+	unsigned int idx = 0;
+	size_t length = 0;
+	size_t dst_rlen = 0;
+	size_t copy_size = 0;
+
+	if (!obj || !obj->orig.data || !obj->priv)
+		return 0;
+
+	dmaobj_cache_operation(TEE_CACHEINVALIDATE, obj);
+
+	priv = obj->priv;
+	sgtdata = priv->sgtdata;
+
+	dst_rlen = obj->orig.length;
+	DMAOBJ_TRACE("Copy (len=%zu)", dst_rlen);
+
+	for (; idx < obj->sgtbuf.number && sgtdata && dst_rlen;
+	     idx++, sgtdata++) {
+		copy_size = MIN(dst_rlen, sgtdata->length);
+		if (sgtdata->orig != sgtdata->dma && sgtdata->orig) {
+			copy_size = MIN(dst_rlen, sgtdata->length);
+			memcpy(sgtdata->orig, sgtdata->dma, copy_size);
+		}
+
+		length += copy_size;
+		dst_rlen -= copy_size;
+	}
+
+	return length;
+}
+
+size_t caam_dmaobj_copy_ltrim_to_orig(struct caamdmaobj *obj)
+{
+	struct priv_dmaobj *priv = NULL;
+	struct sgtdata *sgtdata = NULL;
+	uint8_t *dst = NULL;
+	size_t off = 0;
+	size_t offset = 0;
+	size_t dst_rlen = 0;
+	size_t copy_size = 0;
+	unsigned int idx = 0;
+	size_t length = 0;
+
+	if (!obj || !obj->orig.data || !obj->priv)
+		return 0;
+
+	dmaobj_cache_operation(TEE_CACHEINVALIDATE, obj);
+
+	priv = obj->priv;
+	sgtdata = priv->sgtdata;
+
+	for (; idx < obj->sgtbuf.number && sgtdata; idx++, sgtdata++) {
+		if (!sgtdata->orig)
+			continue;
+
+		for (offset = 0; offset < sgtdata->length; off++, offset++) {
+			if (sgtdata->dma[offset])
+				goto do_copy;
+		}
+	}
+
+do_copy:
+	if (off < obj->orig.length)
+		dst_rlen = obj->orig.length - off;
+
+	dst = obj->orig.data;
+
+	DMAOBJ_TRACE("Copy/Move Offset=%zu (len=%zu) TYPE=%d", off, dst_rlen,
+		     obj->sgtbuf.sgt_type);
+
+	if (!dst_rlen) {
+		dst[0] = 0;
+		return 1;
+	}
+
+	for (; idx < obj->sgtbuf.number && sgtdata && dst_rlen;
+	     idx++, sgtdata++) {
+		if (!sgtdata->orig)
+			continue;
+
+		if (offset) {
+			copy_size = MIN(dst_rlen, sgtdata->length - offset);
+			memmove(dst, &sgtdata->dma[offset], copy_size);
+			offset = 0;
+		} else {
+			copy_size = MIN(dst_rlen, sgtdata->length);
+			if (dst != sgtdata->dma)
+				memmove(dst, sgtdata->dma, copy_size);
+		}
+
+		dst += copy_size;
+		dst_rlen -= copy_size;
+		length += copy_size;
+	}
+
+	return length;
+}
+
+void caam_dmaobj_free(struct caamdmaobj *obj)
+{
+	struct priv_dmaobj *priv = NULL;
+	struct dmaentry *entry = NULL;
+	struct dmaentry *next = NULL;
+	uint32_t exceptions = 0;
+
+	if (!obj)
+		return;
+
+	exceptions = cpu_spin_lock_xsave(&memlock);
+	priv = obj->priv;
+	if (!priv)
+		goto end;
+
+	DMAOBJ_TRACE("Free %s object with data @%p of %zu bytes",
+		     priv->type & DMAOBJ_INPUT ? "Input" : "Output",
+		     obj->orig.data, obj->orig.length);
+
+	entry = TAILQ_FIRST(&priv->list);
+	while (entry) {
```

Comment:
The macro `TAILQ_FOREACH_SAFE` should be useful here.

Reply:
Ok